libsel4/tools/generator_rust.py

```diff
@@ -0,0 +1,604 @@
+
+        def render_parameter_name(self, name):
+            """
+            Return a string of Rust code that would be used in a function
+            parameter declaration.
+            """
+            if name == "type":
+                name = "type_"
+            return "%s: %s" % (name, Generator_Rust.translate_type(self.name))
+
+        def pointer(self, mutable):
+            """
+            Return a new Type class representing a pointer to this object.
+            """
+            return Generator_Rust.PointerType(self, self.wordsize, mutable)
+
+        def rust_expression(self, var_name, word_num=0):
+            """
+            Return code for an expression that gets word 'word_num'
+            of this type.
+            """
+            assert word_num == 0
+            return "%s" % var_name
+
+        def double_word_expression(self, var_name, word_num, word_size):
+
+            assert word_num == 0 or word_num == 1
+
+            if word_num == 0:
+                return "{1} as {0}".format(Generator_Rust.TYPES[self.size_bits], var_name)
+            elif word_num == 1:
+                return "{1}.wrapping_shr({2}) as {0}".format(Generator_Rust.TYPES[self.size_bits], var_name,
+                                                             word_size)
+
+    class BitFieldType(Type):
+        """
+        A special 'struct' generated by the bitfield generator
+        """
+
+        def __init__(self, name, size_bits, wordsize):
+            Generator_Rust.Type.__init__(self, name, size_bits, wordsize)
+
+        def rust_expression(self, var_name, word_num=0):
+            return "%s.words[%d]" % (var_name, word_num)
+
+    class CapType(Type):
+        """
+        A type that is just a typedef of seL4_CPtr.
+        """
+
+        def __init__(self, name, wordsize):
+            Generator_Rust.Type.__init__(self, name, wordsize, wordsize)
+
+    class StructType(Type):
+        """
+        A 'struct' definition.
+        """
+
+        def __init__(self, name, size_bits, wordsize):
+            Generator_Rust.Type.__init__(self, name, size_bits, wordsize)
+
+        def rust_expression(self, var_name, word_num, member_name):
+            assert word_num < self.size_bits // self.wordsize
+
+            # Multiword structure.
+            assert self.pass_by_reference()
+            return "(*%s).%s" % (var_name, member_name[word_num])
+
+    class PointerType(Type):
+        """
+        A pointer to a standard type.
+        """
+
+        def __init__(self, base_type, wordsize, mutable):
+            Generator_Rust.Type.__init__(self, base_type.name, wordsize, wordsize)
+            self.base_type = base_type
+            self.mutable = mutable
+
+        def render_parameter_name(self, name):
+            if name == "type":
+                name = "type_"
+            if self.mutable:
+                return "%s: *mut %s" % (name, Generator_Rust.translate_type(self.name))
```

Comment:
Could merge this and avoid some redundancy (it's a bit less readable then, though):
```
                return "%s: *%s %s" % (
                    "type_" if name == "type" else name, 
                    "mut" if self.mutable else "const", 
                    Generator_Rust.translate_type(self.name))
```

If the constructor of the `Type` class does a `self.generator = Generator_Rust` in its constructor, then `Type` and all derived classes can use `self.generator.some_static_method()`. This avoids having absolute class references and we can do this here:
```
                return "%s: *%s %s" % (
                    "type_" if name == "type" else name, 
                    "mut" if self.mutable else "const", 
                    self.generator.translate_type(self.name))
```